traffic_ops/traffic_ops_golang/deliveryservice/autorenewcerts.go

```diff
@@ -0,0 +1,177 @@
+}
+
+type ExpirationSummary struct {
+	LetsEncryptExpirations []DsExpirationInfo
+	SelfSignedExpirations  []DsExpirationInfo
+	OtherExpirations       []DsExpirationInfo
+}
+
+const emailTemplateFile = "/opt/traffic_ops/app/templates/send_mail/autorenewcerts_mail.html"
+
+func RenewCertificates(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	if inf.Config.RiakEnabled == false {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, errors.New("the Riak service is unavailable"), errors.New("getting SSL keys from Riak by xml id: Riak is not configured"))
+		return
+	}
+
+	rows, err := inf.Tx.Tx.Query(`SELECT xml_id, ssl_key_version FROM deliveryservice`)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, err)
+		return
+	}
+	defer rows.Close()
+
+	keysFound := ExpirationSummary{}
+	for rows.Next() {
+		ds := DsKey{}
+		err := rows.Scan(&ds.XmlId, &ds.Version)
+		if err != nil {
+			log.Errorf("getting delivery services: %v", err)
+			continue
+		}
+		if ds.Version.Valid && ds.Version.Int64 != 0 {
+			continue
+		}
+
+		dsExpInfo := DsExpirationInfo{}
+		keyObj, ok, err := riaksvc.GetDeliveryServiceSSLKeysObj(ds.XmlId, strconv.Itoa(int(ds.Version.Int64)), inf.Tx.Tx, inf.Config.RiakAuthOptions, inf.Config.RiakPort)
+		if err != nil {
+			log.Errorf("getting ssl keys for xmlId: " + ds.XmlId + " and version: " + strconv.Itoa(int(ds.Version.Int64)) + " :" + err.Error())
+			dsExpInfo.XmlId = ds.XmlId
+			dsExpInfo.Version = util.JSONIntStr(int(ds.Version.Int64))
+			dsExpInfo.Error = errors.New("getting ssl keys for xmlId: " + ds.XmlId + " and version: " + strconv.Itoa(int(ds.Version.Int64)) + " :" + err.Error())
+			continue
+		}
+		if !ok {
+			log.Errorf("no object found for the specified key with xmlId: " + ds.XmlId + " and version: " + strconv.Itoa(int(ds.Version.Int64)))
+			dsExpInfo.XmlId = ds.XmlId
+			dsExpInfo.Version = util.JSONIntStr(int(ds.Version.Int64))
+			dsExpInfo.Error = errors.New("no object found for the specified key with xmlId: " + ds.XmlId + " and version: " + strconv.Itoa(int(ds.Version.Int64)))
+			continue
+		}
+
+		err = base64DecodeCertificate(&keyObj.Certificate)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting SSL keys for XMLID '"+ds.XmlId+"': "+err.Error()))
+			return
+		}
+
+		expiration, err := parseExpirationFromCert([]byte(keyObj.Certificate.Crt))
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New(ds.XmlId+": "+err.Error()))
+			return
+		}
+
+		// Renew only certificates within configured limit plus 3 days
+		if expiration.After(time.Now().Add(time.Hour * 24 * time.Duration(inf.Config.ConfigLetsEncrypt.RenewDaysBeforeExpiration)).Add(time.Hour * 24 * 3)) {
```

Comment:
The documentation says that `lets_encrypt.renew_days_before_expiration` sets the "Number of days before expiration date to renew certificates," but this looks like what it actually does is it sets the "number of days before three days before expiration date to renew certificates" - this should either honor the configured value or the configuration documentation should tell the truth about when certificates will be renewed.

Reply:
If I get a vote it's that you make the number of days an unsigned integer and just honor the configuration. Much less confusing.

Reply:
sounds good! im open to changing it to honor the configuration but my thought behind the 3 day bonus is that if something goes wrong in the renewal, then people would still have 3 days to run it again before the cert was expected to be renewed.  if people set a longer time to renew then its not as much of a worry than if people set a short renewal. might be a better option actually to honor the configuration as long as its over 3 days as the safety buffer. whatre your thoughts on that? or just trust people to set a good config for what they need?

Reply:
I think a program trying to be smarter than its own configuration is a recipe for confusion and - potentially - disaster. If you think setting the value too low is dangerous, then the configuration should either disallow setting below the dangerous value (I'd be **extremely** hesitant to do that) or just come with a warning in the documentation.

Reply:
sounds good!

Reply:
done